packages/@aws-cdk/custom-resources/lib/provider-framework/runtime/framework.ts

```diff
@@ -0,0 +1,177 @@
+// tslint:disable: no-console
+// tslint:disable: max-line-length
+import { IsCompleteResponse, OnEventResponse } from '../types';
+import { submitCloudFormationResponse } from './cfn-response';
+import consts = require('./consts');
+import { invokeFunction, startExecution } from './outbound';
+import { failOnError, getEnv, log, Retry } from './util';
+
+// use consts for handler names to compiler-enforce the coupling with construction code.
+export = {
+  [consts.FRAMEWORK_ON_EVENT_HANDLER_NAME]: onEvent,
+  [consts.FRAMEWORK_IS_COMPLETE_HANDLER_NAME]: isComplete,
+  [consts.FRAMEWORK_ON_TIMEOUT_HANDLER_NAME]: onTimeout
+};
+
+/**
+ * The main runtime entrypoint of the async custom resource lambda function.
+ *
+ * Any lifecycle event changes to the custom resources will invoke this handler, which will, in turn,
+ * interact with the user-defined `onEvent` and `isComplete` handlers.
+ *
+ * This function will always succeed. If an error occurs
+ *
+ * @param cfnRequest The cloudformation custom resource event.
+ */
+async function onEvent(cfnRequest: AWSLambda.CloudFormationCustomResourceEvent) {
+  await failOnError(cfnRequest, async () => {
+    log('onEventHandler', cfnRequest);
+
+    cfnRequest.ResourceProperties = cfnRequest.ResourceProperties || { };
+
+    // validate incoming request
+    validateCfnRequest(cfnRequest);
+
+    const onEventResult = await invokeUserFunction(consts.USER_ON_EVENT_FUNCTION_ARN_ENV, cfnRequest) as OnEventResponse;
+    log('onEvent returned:', onEventResult);
+
+    // merge the request and the result from onEvent to form the complete resource event
+    // this also performs validation.
+    const resourceEvent = createResponseEvent(cfnRequest, onEventResult);
+    log('event:', onEventResult);
+
+    // determine if this is an async provider based on whether we have an isComplete handler defined.
+    // if it is not defined, then we are basically ready to return a positive response.
+    if (!process.env[consts.USER_IS_COMPLETE_FUNCTION_ARN_ENV]) {
+      await submitCloudFormationResponse('SUCCESS', resourceEvent);
+      return;
+    }
+
+    // ok, we are not complete, so kick off the waiter workflow
+    const waiter = {
+      stateMachineArn: getEnv(consts.WAITER_STATE_MACHINE_ARN_ENV),
+      name: resourceEvent.RequestId,
+      input: JSON.stringify(resourceEvent),
+    };
+
+    log('starting waiter', waiter);
+
+    // kick off waiter state machine
+    await startExecution(waiter);
+    return;
+  });
+}
+
+// invoked a few times until `complete` is true or until it times out.
+async function isComplete(event: AWSCDKAsyncCustomResource.IsCompleteRequest) {
+  return await failOnError(event, async () => {
```

Comment:
Nit: no need to `await`, can just `return` this. Function doesn't even need to be marked `async`.

Reply:
Same response - I prefer to be a bit more ceramonial with async programming.